File: Coding/golang/graph_algo/graph_algo.go
package main

// Ported from Coding/java/Graph_Algo.java (original package code.ds).
// The original referenced Graph/Vertex/Edge from another file; minimal generic
// versions are defined here so the package is self-contained.

import "fmt"

// Vertex is a graph vertex with an id and adjacency list.
type Vertex[T any] struct {
	id       int64
	adjacent []*Vertex[T]
}

func (v *Vertex[T]) getId() int64                  { return v.id }
func (v *Vertex[T]) getAdjacentVertexes() []*Vertex[T] { return v.adjacent }

// Edge connects two vertices.
type Edge[T any] struct {
	v1 *Vertex[T]
	v2 *Vertex[T]
}

func (e *Edge[T]) getVertex1() *Vertex[T] { return e.v1 }
func (e *Edge[T]) getVertex2() *Vertex[T] { return e.v2 }

// Graph holds all vertices.
type Graph[T any] struct {
	vertices []*Vertex[T]
}

func (g *Graph[T]) getAllVertex() []*Vertex[T] { return g.vertices }

// UndirectedGraphNode is used by the graph-cloning routines.
type UndirectedGraphNode struct {
	label     int
	neighbors []*UndirectedGraphNode
}

func newUndirectedGraphNode(x int) *UndirectedGraphNode {
	return &UndirectedGraphNode{label: x, neighbors: []*UndirectedGraphNode{}}
}

// GraphAlgo groups the graph algorithms (mirrors the Java instance class).
type GraphAlgo struct {
	cloneMap map[int]*UndirectedGraphNode
}

func main() {
	// Build Order via topological sort:
	//   0:        1: 0   2: 0   3: 1,2   4: 3
	dependencies := [][]int{{}, {0}, {0}, {1, 2}, {3}}
	order := buildOrder(dependencies)
	fmt.Println("Build order:", order)
}

// DFS traversal.
func (ga *GraphAlgo) DFS(graph *Graph[int]) {
	visited := map[int64]bool{}
	for _, vertex := range graph.getAllVertex() {
		if !visited[vertex.getId()] {
			ga.DFSUtil(vertex, visited)
		}
	}
}

func (ga *GraphAlgo) DFSUtil(v *Vertex[int], visited map[int64]bool) {
	visited[v.getId()] = true
	fmt.Print(v.getId(), " ")
	for _, vertex := range v.getAdjacentVertexes() {
		if !visited[vertex.getId()] {
			ga.DFSUtil(vertex, visited)
		}
	}
}

// BFS traversal.
func (ga *GraphAlgo) BFS(graph *Graph[int]) {
	visited := map[int64]bool{}
	q := []*Vertex[int]{}
	for _, vertex := range graph.getAllVertex() {
		if !visited[vertex.getId()] {
			q = append(q, vertex)
			visited[vertex.getId()] = true
			for len(q) != 0 {
				vq := q[0]
				q = q[1:]
				fmt.Print(vq.getId(), " ")
				for _, v := range vq.getAdjacentVertexes() {
					if !visited[v.getId()] {
						q = append(q, v)
						visited[v.getId()] = true
					}
				}
			}
		}
	}
}

// hasCycle detects a cycle in a directed graph (white/gray/black sets).
func (ga *GraphAlgo) hasCycle(graph *Graph[int]) bool {
	whiteSet := map[*Vertex[int]]bool{}
	graySet := map[*Vertex[int]]bool{}
	blackSet := map[*Vertex[int]]bool{}

	for _, vertex := range graph.getAllVertex() {
		whiteSet[vertex] = true
	}

	for len(whiteSet) > 0 {
		var current *Vertex[int]
		for v := range whiteSet {
			current = v
			break
		}
		if ga.dfs(current, whiteSet, graySet, blackSet) {
			return true
		}
	}
	return false
}

func (ga *GraphAlgo) dfs(current *Vertex[int], whiteSet, graySet, blackSet map[*Vertex[int]]bool) bool {
	// move current to gray set from white set and then explore it.
	ga.moveVertex(current, whiteSet, graySet)
	for _, neighbor := range current.getAdjacentVertexes() {
		// if in black set means already explored so continue.
		if blackSet[neighbor] {
			continue
		}
		// if in gray set then cycle found.
		if graySet[neighbor] {
			return true
		}
		if ga.dfs(neighbor, whiteSet, graySet, blackSet) {
			return true
		}
	}
	// move vertex from gray set to black set when done exploring.
	ga.moveVertex(current, graySet, blackSet)
	return false
}

func (ga *GraphAlgo) moveVertex(vertex *Vertex[int], sourceSet, destinationSet map[*Vertex[int]]bool) {
	delete(sourceSet, vertex)
	destinationSet[vertex] = true
}

// Find single source shortest path using Dijkstra's algorithm helper.
func (ga *GraphAlgo) getVertexForEdge(v *Vertex[int], e *Edge[int]) *Vertex[int] {
	if e.getVertex1() == v {
		return e.getVertex2()
	}
	return e.getVertex1()
}

// buildOrder returns a valid build order using topological sort. The index is
// the process number; the value is the list of processes it depends on.
func buildOrder(dependencies [][]int) []int {
	temporaryMarks := map[int]bool{}
	permanentMarks := map[int]bool{}
	result := []int{}
	for i := 0; i < len(dependencies); i++ {
		if !permanentMarks[i] {
			result = visit(i, dependencies, temporaryMarks, permanentMarks, result)
		}
	}
	return result
}

// visit searches all unmarked nodes accessible from project.
func visit(project int, dependencies [][]int, temporaryMarks, permanentMarks map[int]bool, result []int) []int {
	// Throw an error if we find a cycle.
	if temporaryMarks[project] {
		panic("Graph is not acyclic")
	}

	if !permanentMarks[project] {
		temporaryMarks[project] = true
		for _, i := range dependencies[project] {
			result = visit(i, dependencies, temporaryMarks, permanentMarks, result)
		}
		permanentMarks[project] = true
		delete(temporaryMarks, project)
		result = append(result, project)
	}
	return result
}

// clone clones a graph when node labels are unique.
func (ga *GraphAlgo) clone(node *UndirectedGraphNode) *UndirectedGraphNode {
	if node == nil {
		return nil
	}
	if ga.cloneMap == nil {
		ga.cloneMap = map[int]*UndirectedGraphNode{}
	}
	if v, ok := ga.cloneMap[node.label]; ok {
		return v
	}
	tmp := newUndirectedGraphNode(node.label)
	ga.cloneMap[tmp.label] = tmp
	for _, neighbor := range node.neighbors {
		tmp.neighbors = append(tmp.neighbors, ga.clone(neighbor))
	}
	return tmp
}

// cloneWithVisited clones a graph that may contain duplicate labels.
func cloneWithVisited(src *UndirectedGraphNode, visited map[*UndirectedGraphNode]*UndirectedGraphNode) *UndirectedGraphNode {
	if src == nil {
		return nil
	}
	if v, ok := visited[src]; ok {
		return v
	}
	tmp := newUndirectedGraphNode(src.label)
	visited[src] = tmp
	for _, child := range src.neighbors {
		tmp.neighbors = append(tmp.neighbors, cloneWithVisited(child, visited))
	}
	return tmp
}

File: Coding/python/graph_algo.py
"""Idiomatic Python 3 port of java/Graph_Algo.java (original package code.ds).

Graph algorithms: DFS / BFS traversal, directed-graph cycle detection
(white/gray/black), topological "build order", and cloning of an undirected
graph.

The Java source references ``Vertex`` / ``Edge`` / ``Graph`` from a separate
file; minimal self-contained versions are defined here so the module runs
standalone. Java had two overloaded ``clone`` methods, renamed here to
``clone_unique`` (unique labels, uses an instance map) and
``clone_with_visited`` (possibly duplicate labels, identity-keyed visited map).
The Java source has no ``main``; a demonstrative one is added.
"""

from collections import deque


# ---------------------------------------------------------------------------
# Minimal graph model (Vertex / Edge / Graph)
# ---------------------------------------------------------------------------
class Vertex:
    def __init__(self, vid):
        self.id = vid
        self.adjacent = []

    def get_id(self):
        return self.id

    def get_adjacent_vertexes(self):
        return self.adjacent

    def add_adjacent(self, v):
        self.adjacent.append(v)


class Edge:
    def __init__(self, v1, v2, weight=0):
        self.vertex1 = v1
        self.vertex2 = v2
        self.weight = weight

    def get_vertex1(self):
        return self.vertex1

    def get_vertex2(self):
        return self.vertex2


class Graph:
    def __init__(self, directed=True):
        self.directed = directed
        self.vertices = {}  # id -> Vertex (insertion ordered)

    def _get_or_create(self, vid):
        v = self.vertices.get(vid)
        if v is None:
            v = Vertex(vid)
            self.vertices[vid] = v
        return v

    def add_vertex(self, vid):
        self._get_or_create(vid)

    def add_edge(self, id1, id2):
        v1 = self._get_or_create(id1)
        v2 = self._get_or_create(id2)
        v1.add_adjacent(v2)
        if not self.directed:
            v2.add_adjacent(v1)

    def get_all_vertex(self):
        return list(self.vertices.values())


class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []


# ---------------------------------------------------------------------------
# Graph algorithms
# ---------------------------------------------------------------------------
class GraphAlgo:
    def __init__(self):
        self.map = {}  # label -> node, used by clone_unique

    # --- DFS traversal ---
    def dfs_traversal(self, graph):
        visited = set()
        for vertex in graph.get_all_vertex():
            if vertex.get_id() not in visited:
                self._dfs_util(vertex, visited)

    def _dfs_util(self, v, visited):
        visited.add(v.get_id())
        print(v.get_id(), end=" ")
        for vertex in v.get_adjacent_vertexes():
            if vertex.get_id() not in visited:
                self._dfs_util(vertex, visited)

    # --- BFS traversal ---
    def bfs_traversal(self, graph):
        visited = set()
        q = deque()
        for vertex in graph.get_all_vertex():
            if vertex.get_id() not in visited:
                q.append(vertex)
                visited.add(vertex.get_id())
                while len(q) != 0:
                    vq = q.popleft()
                    print(vq.get_id(), end=" ")
                    for v in vq.get_adjacent_vertexes():
                        if v.get_id() not in visited:
                            q.append(v)
                            visited.add(v.get_id())

    # --- Cycle detection in a directed graph ---
    def has_cycle(self, graph):
        white_set = set(graph.get_all_vertex())
        gray_set = set()
        black_set = set()

        while len(white_set) > 0:
            current = next(iter(white_set))
            if self._dfs_cycle(current, white_set, gray_set, black_set):
                return True
        return False

    def _dfs_cycle(self, current, white_set, gray_set, black_set):
        # Move current to gray (being explored) from white, then explore it.
        self._move_vertex(current, white_set, gray_set)
        for neighbor in current.get_adjacent_vertexes():
            if neighbor in black_set:  # already fully explored
                continue
            if neighbor in gray_set:  # back edge -> cycle
                return True
            if self._dfs_cycle(neighbor, white_set, gray_set, black_set):
                return True
        # Done exploring: move from gray to black.
        self._move_vertex(current, gray_set, black_set)
        return False

    def _move_vertex(self, vertex, source_set, destination_set):
        source_set.discard(vertex)
        destination_set.add(vertex)

    def get_vertex_for_edge(self, v, e):
        return e.get_vertex2() if e.get_vertex1() is v else e.get_vertex1()

    # --- Clone undirected graph (unique labels) ---
    def clone_unique(self, node):
        if node is None:
            return None
        if node.label in self.map:
            return self.map[node.label]
        tmp = UndirectedGraphNode(node.label)
        self.map[tmp.label] = tmp
        for neighbor in node.neighbors:
            tmp.neighbors.append(self.clone_unique(neighbor))
        return tmp

    # --- Clone undirected graph (labels may repeat; identity visited map) ---
    def clone_with_visited(self, src, visited):
        if src is None:
            return None
        if id(src) in visited:
            return visited[id(src)]
        tmp = UndirectedGraphNode(src.label)
        visited[id(src)] = tmp
        for child in src.neighbors:
            tmp.neighbors.append(self.clone_with_visited(child, visited))
        return tmp


# ---------------------------------------------------------------------------
# Build order via topological sort (Java static methods)
# ---------------------------------------------------------------------------
def build_order(dependencies):
    temporary_marks = set()
    permanent_marks = set()
    result = []
    for i in range(len(dependencies)):
        if i not in permanent_marks:
            visit(i, dependencies, temporary_marks, permanent_marks, result)
    return result


def visit(project, dependencies, temporary_marks, permanent_marks, result):
    if project in temporary_marks:
        raise RuntimeError("Graph is not acyclic")
    if project not in permanent_marks:
        temporary_marks.add(project)
        for i in dependencies[project]:
            visit(i, dependencies, temporary_marks, permanent_marks, result)
        permanent_marks.add(project)
        temporary_marks.discard(project)
        result.append(project)


if __name__ == "__main__":
    algo = GraphAlgo()
    g = Graph(directed=True)
    for a, b in [(0, 1), (0, 2), (1, 3), (2, 3), (3, 4)]:
        g.add_edge(a, b)

    print("DFS:", end=" ")
    algo.dfs_traversal(g)
    print()

    print("BFS:", end=" ")
    algo.bfs_traversal(g)
    print()

    print("has_cycle:", algo.has_cycle(g))

    deps = [[], [0], [0], [1, 2], [3]]
    print("Build order:", build_order(deps))

    # clone demo (unique labels)
    n0 = UndirectedGraphNode(0)
    n1 = UndirectedGraphNode(1)
    n2 = UndirectedGraphNode(2)
    n0.neighbors = [n1, n2]
    n1.neighbors = [n2]
    cloned = GraphAlgo().clone_unique(n0)
    print("clone label:", cloned.label, "neighbors:", [n.label for n in cloned.neighbors])
